// Copyright 2019 Zeppelin Bend Pty Ltd
// This file is part of cimbend.
// 
// cimbend is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// cimbend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with cimbend.  If not, see <https://www.gnu.org/licenses/>.


syntax = "proto3";
option java_multiple_files = true;
option java_package = "com.zepben.protobuf.cim.iec61970.base.wires";
package zepben.cim.iec61970.base.wires;

import "zepben/cim/iec61968/common/Location.proto";
import "zepben/cim/iec61970/base/core/Terminal.proto";
import "zepben/cim/iec61970/base/diagramlayout/DiagramObject.proto";
import "zepben/cim/iec61970/base/meas/Control.proto";
import "zepben/cim/iec61970/base/meas/Measurement.proto";
import "zepben/cim/iec61970/base/wires/PowerTransformerEnd.proto";
import "zepben/cim/iec61970/base/wires/VectorGroup.proto";

/**
 * An electrical device consisting of  two or more coupled windings, with or without a magnetic core, for introducing mutual coupling
 * between electric circuits. Transformers can be used to control voltage and phase shift (active power flow).
 * <p>
 * A power transformer may be composed of separate transformer tanks that need not be identical.
 * <p>
 * A power transformer can be modeled with or without tanks and is intended for use in both balanced and unbalanced representations. A
 * power transformer typically has two terminals, but may have one (grounding), three or more terminals.
 * <p>
 * The inherited association ConductingEquipment.BaseVoltage should not be used.  The association from TransformerEnd to BaseVoltage
 * should be used instead.
 */
message PowerTransformer {

    /**
     * Master resource identifier issued by a model authority. The mRID is unique within an exchange context. Global uniqueness
     * is easily achieved by using a UUID,  as specified in RFC 4122, for the mRID. The use of UUID is strongly recommended.
     * <p>
     * For CIMXML data files in RDF syntax conforming to IEC 61970-552 Edition 1, the mRID is mapped to rdf:ID or rdf:about attributes
     * that identify CIM object elements.
     */
    string mRID = 1;

    /**
     * The name is any free human readable and possibly non unique text naming the object.
     */
    string name = 2;

    /**
     * If true, the equipment is in service.
     */
    bool inService = 3;

    /**
     * If true, the equipment is normally in service.
     */
    bool normallyInService = 4;

    /**
     * Base voltage of this conducting equipment.  Use only when there is no voltage level container used and only one base
     * voltage applies.  For example, not used for transformers.
     */
    string baseVoltageMRID = 5;

    /**
     * Conducting equipment have terminals that may be connected to other conducting equipment terminals via connectivity nodes
     * or topological nodes.
     * Note that the order of the terminals is important, and reflects the relationship to PowerTransformerEnds,
     * and PositionPoints.
     */
    repeated iec61970.base.core.Terminal terminals = 6;

    /**
     * The PowerTransformerEnd's for this PowerTransformer. Each end is associated with a terminal of the
     * PowerTransformer, and the order of the ends reflects the Terminal the end is associated with.
     * For example, end[0] is the end for terminal[0].
     */
    repeated iec61970.base.wires.PowerTransformerEnd powerTransformerEnds = 7;

    /**
     * Vector group of the transformer for protective relaying. For unbalanced transformers, this may not be simply
     * determined from the constituent winding connections and phase angle dispacements.
     */
    iec61970.base.wires.VectorGroup vectorGroup = 8;

    /**
     * The diagram objects that are associated with the domain object.
     * <p>
     * The diagram objects are returned in a map sorted by the diagram they live in.
     */
    repeated iec61970.base.diagramlayout.DiagramObject diagramObjects = 16;

    /**
     * Datasheet information for this power system resource.
     */
    string assetInfoMRID = 17;

    /**
     * Location of this power system resource.
     */
    iec61968.common.Location location = 18;

    /**
     * The controller outputs used to actually govern a regulating device, e.g. the magnetization of a synchronous machine or
     * capacitor bank breaker actuator.
     */
    iec61970.base.meas.Control control = 19;

    /**
     * The measurements associated with this power system resource.
     */
    iec61970.base.meas.Measurement measurement = 20;

}
